src/kbucket.rs

```diff
@@ -1,219 +1,87 @@
-#![allow(unused)]
-
-use crate::helper::{Identifier, Node, U256};
+use crate::helper::{Identifier, U256};
 use std::collections::HashMap;
 use std::net::Ipv4Addr;
-use uint::*;
 
 const BUCKET_SIZE: usize = 20;
 const MAX_BUCKETS: usize = 256;
 
-type Bucket = [Option<Node>; BUCKET_SIZE];
+// Should I create a HashMap with a generic K, V or have my hashmap take in an id and TR?
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Bucket {
+    pub map: HashMap<Identifier, TableRecord>,
+    pub limit: usize,
+}
+
+impl Bucket {
+    fn new(&self, limit: usize) -> Self {
+        Bucket {
+            map: HashMap::new(),
+            limit,
+        }
+    }
 
-#[derive(Debug)]
-pub enum FindNodeResult {
-    Found(Option<Node>),
-    NotFound(Vec<Option<Node>>),
+    fn insert(&mut self, key: Identifier, value: TableRecord) -> Option<TableRecord> {
+        if self.map.len() <= BUCKET_SIZE {
+            self.map.insert(key, value)
+        } else {
+            None
+        }
+    }
 }
 
-#[derive(Debug)]
-enum Search {
-    Success(usize, usize),
-    Failure(usize, usize),
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct TableRecord {
+    pub node_id: Identifier,
+    pub ip_address: Ipv4Addr,
+    pub udp_port: u16,
 }
 
 // Bucket 0: Closest peers from node in network.
 // Bucket 255: Farthest peers from node in network
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
 pub struct KbucketTable {
     pub local_node_id: Identifier,
-    pub buckets: [Bucket; MAX_BUCKETS],
-    store: HashMap<Vec<u8>, Vec<u8>>, // Same storage as portal network.
+    pub local_record: TableRecord,
+    pub buckets: Vec<Bucket>,
 }
 
 impl KbucketTable {
-    pub fn new(local_node_id: Identifier) -> Self {
+    pub fn new(local_node_id: Identifier, local_record: TableRecord) -> Self {
```

Comment:
see above comment, you are already passing `local_node_id` in the `local_record` so I'd just take one or the other